### PR TITLE
Scratchpad Entity

### PR DESCRIPTION
**Pull request by gbuisson**
Add CTIM Scratchpad entity support as well as a new `PATCH` flow intended for partial updates
- [x] Store
- [x] Basic CRUD routes
- [x] Atomic operation routes
- [x] Graphql
- [x] migration support

Misc: 
- added support to migrate `Investigation`entities
- added `abstraction_level` to `AttackPattern`entity mapping 
- Fixed Swagger page title

### CODE REVIEW

File: test/ctia/http/routes/scratchpad_test.clj [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+      (testing "POST /ctia/scratchpad/:id/observables :add"
+        (let [new-observables [{:type "ip" :value "42.42.42.42"}]
+              response (post (str "ctia/scratchpad/" (:short-id scratchpad-id) "/observables")
+                             :body {:operation :add
+                                    :observables new-observables}
+                             :headers {"Authorization" "45c1f5e3f05d0"})
+              updated-scratchpad (:parsed-body response)]
+          (is (= 200 (:status response)))
+          (is (subset? (set new-observables)
+                       (set (:observables updated-scratchpad))))))
```
Comment (msprunck) [inline test/ctia/http/routes/scratchpad_test.clj]: I suggest adding a test to check that `new-observables` is not equal to `(:observables updated-scratchpad)`.

Same comment for all `:add` operations on bundle entities and texts.

File: src/ctia/flows/crud.clj [hunk 2]
```diff
@@ -296,6 +300,49 @@
...
+                   :remove remove-colls
+                   :replace replace-colls
+                   replace-colls)
+        entity (-> (deep-merge-with patch-fn
+                                    prev-entity
+                                    partial-entity)
+                   un-store
+                   (dissoc :id))]
+    (assoc fm :entities [entity])))
+
```
Comment (yogsototh) [inline src/ctia/flows/crud.clj]: All those functions are really "data manipulation" about things without types or form.
Its hard to follow. Also I couldn't find any test showing how to use them and what to expect from them individually.

Would you mind to add a docstring with example and if possible unit test for each of these functions with simple examples? Typically for `add-colls`, `remove-colls` and `replace-colls`.

File: src/ctia/http/routes/common.clj [hunk 3]
```diff
@@ -398,6 +401,28 @@
...
+  {(s/optional-key :fields) [scratchpad-sort-fields]})
+
+(s/defschema ScratchpadSearchParams
+  (st/merge
+   PagingParams
+   BaseEntityFilterParams
+   SourcableEntityFilterParams
+   ScratchpadFieldsParam
+   {:query s/Str
+    (s/optional-key :texts.text) s/Str
```
Comment (yogsototh) [inline src/ctia/http/routes/common.clj]: I know that certainly "works™" right now. But this is "dangerous" from:

https://clojure.org/reference/reader

> Keywords - Keywords are like symbols, except:
> ... they cannot contain '.' or name classes.

So the solution might be to use strings instead of keywords for query if possible and easy to do. Certainly something to point in another technical issue. I imagine the problem is the same for all other entities.

File: test/ctia/http/generative/es_store_spec.clj [hunk 4]
```diff
@@ -56,3 +56,6 @@
 
 (defspec ^:generative api-for-tool-routes-es-store
   prop/api-for-tool-routes)
+
+#_(defspec ^:disabled api-for-scratchpad-routes-es-store
+    prop/api-for-scratchpad-routes)
```
Comment (yogsototh) [inline test/ctia/http/generative/es_store_spec.clj]: You should either remove it before merging or use a `;; TODO: fix it` comment + file an issue not to forget it.

File: test/ctia/http/generative/properties.clj [hunk 5]
```diff
@@ -136,3 +142,7 @@
 (def api-for-tool-routes
   (api-for-route 'tool
                  (spec-gen "max-new-tool")))
+
+#_(def api-for-scratchpad-routes
```
Comment (yogsototh) [inline test/ctia/http/generative/properties.clj]: same here, see previous comment.

File: test/ctia/http/routes/actor_test.clj [hunk 6]
```diff
@@ -92,20 +76,7 @@
...
-                :actor_type "Hacker",
-                :title "actor",
-                :confidence "High",
-                :source "a source"
-                :valid_time {:start_time #inst "2016-02-11T00:40:48.212-00:00"
-                             :end_time #inst "2016-07-11T00:40:48.212-00:00"}
-                :schema_version schema-version
-                :tlp "green"}
-               actor))))
+               (assoc new-actor :id (id/long-id actor-id)) actor))))
```
Comment (yogsototh) [inline test/ctia/http/routes/actor_test.clj]: so much better!

File: test/ctia/http/routes/scratchpad_test.clj [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+                                            :name "TEST"
+                                            :labels ["malware"]}}}
+              response (post (str "ctia/scratchpad/" (:short-id scratchpad-id) "/bundle")
+                             :body {:operation :replace
+                                    :bundle bundle-entities}
+                             :headers {"Authorization" "45c1f5e3f05d0"})
+              updated-scratchpad (:parsed-body response)]
+          (is (= 200 (:status response)))
+          (is (= (:malwares bundle-entities)
+                 (-> updated-scratchpad :bundle :malwares)))))
```
Comment (yogsototh) [inline test/ctia/http/routes/scratchpad_test.clj]: I suggest to add a test that check that only the `:malwares` value was updated.
For example using a GET just before.

~~~
(is (= old-bundle (into old-bundle bundle-entities)))
~~~

I'm not sure how :replace could take advantage of `deep-merging` instead of a simple `into` there. It would be nice to have an example for that specific case that will also help to test that we only replace a specific sub element of the scratchpad.

**Comment by gbuisson** [comment]
I adressed all your feedback, also released `clj-momo` (I forgot about it ;-))

**Comment by gbuisson** [comment]
@msprunck you do that with any of the dedicated routes:

- adding a CTIM Object is about patching the bundle to add an entity so you should use POST /scratchpad/scratchpad-foo/bundle with the entity you wish to add

- a CTIM object id is in fact a Bundle reference, like `actor_refs` for instance you can use the same route is above.

**Comment by hshahman** [comment]
verified in rel-v0.3
